From 13838fa7362fc6dcb02b5f27778ac0987bc4083a Mon Sep 17 00:00:00 2001
From: Zi Shen Lim <zlim@netlogicmicro.com>
Date: Wed, 4 Jan 2012 16:23:54 -0800
Subject: [PATCH 320/565] Fix TLB flushing issue

Stale tlbsize was used resulting in incomplete TLB flushes.
During mmu_init, make sure we update tlbsize *after* configuration.

Based on Broadcom SDK 2.3.

Signed-off-by: Zi Shen Lim <zlim@netlogicmicro.com>
Signed-off-by: Yanjiang Jin <yanjiang.jin@windriver.com>
---
 arch/mips/netlogic/xlp/mmu.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/arch/mips/netlogic/xlp/mmu.c b/arch/mips/netlogic/xlp/mmu.c
index a654e25..4aac3fa 100644
--- a/arch/mips/netlogic/xlp/mmu.c
+++ b/arch/mips/netlogic/xlp/mmu.c
@@ -296,11 +296,6 @@ void mmu_init(void)
 	pgwalker_workaround_setup();
 
 	/*
-	 * Read back TLB entries after configuration
-	 */
-	current_cpu_data.tlbsize = (read_c0_config6() >> 16 ) & 0xffff;
-
-	/*
 	 * shift right half the number of 1s in
 	 * the pagemask and populate that value
 	 */
@@ -337,4 +332,10 @@ void mmu_init(void)
 
 	/* Intialize after pgwalker and others are configured! */
 	write_c0_config6(read_c0_config6() | tlb_config);
+
+	/*
+	 * Read back TLB entries after configuration
+	 */
+	current_cpu_data.tlbsize = ((read_c0_config6() >> 16) & 0xffff) + 1;
 }
+
-- 
1.8.4.93.g57e4c17

